instaloader/nodeiterator.py

```diff
@@ -129,7 +129,9 @@ def __next__(self) -> T:
                 raise
             item = self._node_wrapper(node)
             if self._first_node is None:
-                self._first_node = node
+                pinned = 'pinned_for_users' in node and bool(node['pinned_for_users'])
```

Comment:
I would prefer that this condition becomes a parameter of the class, such that the NodeIterator implementation does not need to know anything about the structure it is yielding.

Reply:
I introduced a callback as a parameter to the class constructor, if passed this callback is used to determine what constitutes the first item.

As a bonus, the problem that could lead to redownloads (described in https://github.com/instaloader/instaloader/pull/1586#issuecomment-1157674263) was fixed: instead of skipping pinned posts and storing the time of the first non-pinned post, the callback now can identify the newest post in the feed, pinned or not, even if it's not the first that instagram returns.